### PR TITLE
feat(segmenter): add location for segmenters

### PR DESCRIPTION
**Pull request by shakurshams**

### CODE REVIEW

**Review by Yongxuanzhang — COMMENTED** [review]
Hi, to pass the CI, you need to add the version of each modified segmenters like this:
https://github.com/jina-ai/jina-hub/blob/803453070c4b636c224175cb70297e1c3d767715/segmenters/audio/AudioSlicer/manifest.yml#L10
change it to 0.0.12

File: segmenters/image/FaceNetSegmenter/__init__.py [hunk 1]
```diff
@@ -82,20 +81,30 @@ def segment(self, data: 'np.ndarray', *args, **kwargs) -> List[Dict]:
...
+                faces = faces[0].view(-1, image.shape[-1], self.image_size, self.image_size)
+                batch_boxes = batch_boxes[0]
+                batch_probs = batch_probs[0]
 
             results = [
                 dict(
                     offset=0,
                     weight=probability,
-                    blob=face.detach().numpy(),
+                    blob=face.numpy(),
```
Comment (JoanFM) [inline segmenters/image/FaceNetSegmenter/__init__.py:104]: are u sure about removing `detach`?

Comment (shakurshams) [reply segmenters/image/FaceNetSegmenter/__init__.py:104]: yeah because using `with torch.no_grad()` context manager.

Comment (shakurshams) [reply segmenters/image/FaceNetSegmenter/__init__.py:104]: btw followed the implementation from the `forward` function in their [library](https://github.com/timesler/facenet-pytorch/blob/master/models/mtcnn.py#L226) which uses bounding box and face extraction api internally .

File: segmenters/audio/AudioSlicer/__init__.py [hunk 2]
```diff
@@ -24,15 +23,17 @@ def _segment(self, signal):
...
             frames = librosa.util.frame(signal, frame_length=self.frame_length, hop_length=self.hop_length, axis=0)
+            return frames,
+
         elif signal.ndim == 2:  # stereo
             left_frames = librosa.util.frame(
                 signal[0,], frame_length=self.frame_length, hop_length=self.hop_length, axis=0)
             right_frames = librosa.util.frame(
                 signal[1,], frame_length=self.frame_length, hop_length=self.hop_length, axis=0)
-            frames = np.concatenate((left_frames, right_frames), axis=0)
+            return left_frames, right_frames
```
Comment (JoanFM) [inline segmenters/audio/AudioSlicer/__init__.py:33]: what is the reasoning behind this change?

File: segmenters/image/FaceNetSegmenter/__init__.py [hunk 3]
```diff
@@ -82,20 +81,30 @@ def segment(self, data: 'np.ndarray', *args, **kwargs) -> List[Dict]:
             # Create a batch of size 1
             image = image.unsqueeze(0)
 
-            faces, probabilities = self.face_detector(image, return_prob=True)
+            # Detect faces
+            batch_boxes, batch_probs, batch_points = self.face_detector.detect(image, landmarks=True)
```
Comment (JoanFM) [inline segmenters/image/FaceNetSegmenter/__init__.py:85]: what are the keypoints used for?

**Comment by JoanFM** [comment]
recheckcla

File: segmenters/audio/AudioSlicer/__init__.py (hunk 2 above)
Comment (nan-wang) [reply segmenters/audio/AudioSlicer/__init__.py:33]: I've got the same question here. The only difference is to convert the return from a `ndarray` to a tuple of `ndarray`

Comment (nan-wang) [reply segmenters/audio/AudioSlicer/__init__.py:33]: At the end, both left and right frames are considered the same chunks.

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: segmenters/image/FaceNetSegmenter/__init__.py (hunk 3 above)
Comment (shakurshams) [reply segmenters/image/FaceNetSegmenter/__init__.py:85]: they are not being used anywhere at the moment. the same api `mtcnn.detect()`  returns bounding boxes, probabilities and landmarks if `landmark` is set true. If no future use, we can get rid of these but they get calculated in the library api anyway.

File: segmenters/audio/AudioSlicer/__init__.py (hunk 2 above)
Comment (shakurshams) [reply segmenters/audio/AudioSlicer/__init__.py:33]: It differs in how segmented chunks are stored and reconstructed later if needed.

For `mono` it is same as before
`[ {'offset':` 0, 'location': [0, 10]}, {'offset': 1, 'location': [10, 20]}, {'offset': 2, 'location': [20, 30]} ]`

But for stereo it is different. For each channel in the audio offset starts from 0. This is done keeping reconstruction in mind.
When you see `offset=0` for the second time while traversing the list, you know the chunks for the second channel has started
`[ {'offset': 0, 'location': [0, 10]}, {'offset': 1, 'location': [10, 20]}, {'offset': 2, 'location': [20, 30]},
    {'offset': 0, 'location': [0, 10]}, {'offset': 1, 'location': [10, 20]}, {'offset': 2, 'location': [20, 30]} ]`
    
So when concatenated array is returned from the `_segment` there is no way to understand where the second channel frames start from because we do not have length information. That is why returning a tuple of frames in both cases of mono and stereo and looping over them. Thus no length information is required.

**Comment by shakurshams** [comment]
**_I have read the CLA Document and I hereby sign the CLA_**

File: segmenters/image/FaceNetSegmenter/__init__.py (hunk 3 above)
Comment (JoanFM) [reply segmenters/image/FaceNetSegmenter/__init__.py:85]: If they are not used, let's remove them. Code looks more confusing with unused variables

Comment (shakurshams) [reply segmenters/image/FaceNetSegmenter/__init__.py]: Subsequent box selection method [select_boxes()](https://github.com/timesler/facenet-pytorch/blob/master/models/mtcnn.py#L357) requires the landmarks despite the first method [detect()](https://github.com/timesler/facenet-pytorch/blob/master/models/mtcnn.py#L272) was called with `landmarks=False`. In the [forward()](https://github.com/timesler/facenet-pytorch/blob/master/models/mtcnn.py#L226) method they calculate the landmarks anyway [here](https://github.com/timesler/facenet-pytorch/blob/master/models/mtcnn.py#L258) but just does not return it. Not a clean api design.
I have converted the unused variable to `_`

File: segmenters/audio/AudioSlicer/__init__.py (hunk 2 above)
Comment (shakurshams) [reply segmenters/audio/AudioSlicer/__init__.py:33]: @nan-wang @JoanFM what do you think?

Comment (nan-wang) [reply segmenters/audio/AudioSlicer/__init__.py:33]: @shakurshams Yes, I see your point. However, when you assign the chunks at L53, the frames from the right channel will override those from the left.

**Review by nan-wang — CHANGES_REQUESTED** [review]
Would you please resolve the conflicts? We've fixed the typos in the manifest

File: segmenters/audio/AudioSlicer/__init__.py (hunk 2 above)
Comment (shakurshams) [reply segmenters/audio/AudioSlicer/__init__.py:33]: @nan-wang I have pasted the output from the last two test cases(for one signal instead of two for brevity.) For stereo, chunks from both channels are preserved following the scheme described above. Please let me know if I missed your point.
**Stereo**
`[[{'offset': 0, 'weight': 1.0, 'blob': array([ 0.52708615, -0.21221454, -1.0650824 , -0.48549198, -0.08167043,
       -0.38034069,  1.84052205,  1.26871832, -0.683173  ,  0.21671767]), 'location': [0, 10]}, {'offset': 1, 'weight': 1.0, 'blob': array([-1.55060346, -2.26714673,  0.63778872,  0.86921508,  1.43830099,
       -1.75240552,  2.09746676,  0.77696722,  1.74378927,  0.2115113 ]), 'location': [10, 20]}, {'offset': 2, 'weight': 1.0, 'blob': array([-1.65390314, -0.04860204,  0.03966604, -0.20333415,  1.57110702,
        2.7834961 ,  0.51260664,  0.89482025,  0.10671443,  0.69157241]), 'location': [20, 30]}, {'offset': 3, 'weight': 1.0, 'blob': array([ 0.01885564, -0.8876197 ,  0.67363343, -0.37729471, -0.129511  ,
       -0.93622562, -0.16900971, -1.07119272,  0.85680062, -2.29911036]), 'location': [30, 40]}, {'offset': 4, 'weight': 1.0, 'blob': array([ 0.35992368, -0.04810679, -2.1906127 ,  1.06116999,  0.78800577,
        0.62752204,  0.86104441, -1.46333466,  2.11331119,  1.66678649]), 'location': [40, 50]}, {'offset': 0, 'weight': 1.0, 'blob': array([ 0.23663368, -2.04901163, -1.26061622,  0.00849126, -0.80775948,
        0.32778837,  0.47677929,  0.61443827, -0.83340377, -0.0171386 ]), 'location': [0, 10]}, {'offset': 1, 'weight': 1.0, 'blob': array([-0.39726652,  0.40275055,  1.06923626,  0.46275345, -2.01034733,
       -0.72624474,  0.7360374 , -0.52888659,  0.93563908,  0.88169216]), 'location': [10, 20]}, {'offset': 2, 'weight': 1.0, 'blob': array([ 1.92583051, -1.96131975,  0.9781112 , -0.70289831, -0.46250962,
        1.1666676 ,  0.69058627,  1.53404274,  0.28370558,  0.66857541]), 'location': [20, 30]}, {'offset': 3, 'weight': 1.0, 'blob': array([-1.18176242,  0.87834701,  0.83459   , -2.63591185, -0.57192406,
       -0.88206868,  0.79404908,  1.1386616 , -0.46092391, -0.05199123]), 'location': [30, 40]}, {'offset': 4, 'weight': 1.0, 'blob': array([ 1.06539497, -1.33079596,  0.89450982,  1.9121258 ,  0.14376057,
       -0.35467981, -0.37416159, -1.54486295, -1.27437986,  0.08780342]), 'location': [40, 50]}]]`
       
**Mono:**
`[[{'offset': 0, 'weight': 1.0, 'blob': array([-0.45984967, -0.63631383, -0.66508786, -0.06480296, -1.65522152,
       -0.63912864,  0.14825386,  0.72792373,  0.22836782,  0.03122948]), 'location': [0, 10]}, {'offset': 1, 'weight': 1.0, 'blob': array([-0.3267693 , -0.23741972, -1.30259229, -1.12866133, -0.07516026,
        0.7373738 , -1.63273306, -0.65924093,  0.11204031,  1.0211533 ]), 'location': [10, 20]}, {'offset': 2, 'weight': 1.0, 'blob': array([-0.75463424,  0.13538528,  1.25098552,  0.83004809, -0.31403187,
        0.57925337, -0.79246062, -1.25144412,  1.35942729, -0.4936783 ]), 'location': [20, 30]}, {'offset': 3, 'weight': 1.0, 'blob': array([ 0.49459769, -0.23233242, -1.56765216, -0.57442509,  0.81551088,
       -1.00448579, -0.94726661, -0.67283012, -0.71853731, -1.54388408]), 'location': [30, 40]}, {'offset': 4, 'weight': 1.0, 'blob': array([ 0.18335447,  0.36723736,  1.50778129,  0.02037247, -1.12726425,
       -0.52474678,  1.95070634,  0.21581817, -0.17478254,  0.55028923]), 'location': [40, 50]}]]`

**Comment by shakurshams** [comment]
> Would you please resolve the conflicts? We've fixed the typos in the manifest

Resolved and rebased to upstream master. Tests are passing.

File: segmenters/audio/AudioSlicer/tests/test_audioslicer.py [hunk 4]
```diff
@@ -26,3 +26,31 @@ def test_slice_stereo():
...
+    frame_length = 10
+    signal_orig = np.random.randn(2, frame_length * n_frames)
+    expected_locations = [[0, 10], [10, 20], [20, 30], [30, 40], [40, 50]]
+
+    segmenter = AudioSlicer(frame_length, frame_length)
+
+    segmented_chunks_per_doc = segmenter.segment(np.stack([signal_orig, signal_orig]))
+    for segmented_chunk in segmented_chunks_per_doc:
+        for i, chunk in enumerate(segmented_chunk):
+            assert chunk['location'] == expected_locations[i % n_frames]
```
Comment (nan-wang) [inline segmenters/audio/AudioSlicer/tests/test_audioslicer.py:56]:
```suggestion
def test_location_stereo():
    frame_length = 10
    hop_length = 5
    n_frames = 5
    num_docs = 3
    num_channels = 2
    signal_orig = np.random.randn(num_channels, frame_length * n_frames)
    expected_n_frames = (signal_orig.shape[1] - frame_length) / hop_length
    expected_locations = [[i * frame_length, i * frame_length + frame_length] for i in range(int(expected_n_frames))]

    segmenter = AudioSlicer(frame_length=frame_length, hop_length=frame_length)

    docs = segmenter.segment(np.stack([signal_orig] * num_docs))
    assert len(docs) == num_docs
    for d in docs:
        assert len(d) == n_frames * num_channels
        for i, chunk in enumerate(d):
            assert chunk['location'] == expected_locations[i % n_frames]
```

Comment (nan-wang) [reply segmenters/audio/AudioSlicer/tests/test_audioslicer.py:56]: We can make the test more general and easier to maintain.

File: segmenters/audio/AudioSlicer/__init__.py [hunk 5]
```diff
@@ -43,7 +44,13 @@ def segment(self, blob: 'np.ndarray', *args, **kwargs) -> List[Dict]:
...
+        chunks = []
+        for frames in channel_frames:
+            start = 0
+            for idx, frame in enumerate(frames):
+                chunks.append(dict(offset=idx, weight=1.0, blob=frame, location=[start, start + len(frame)]))
+                start += self.hop_length
 
-        return [dict(offset=idx, weight=1.0, blob=frame)
-                for idx, frame in enumerate(frames)]
+        return chunks
```
Comment (nan-wang) [inline segmenters/audio/AudioSlicer/__init__.py:56]:
```
        channel_frames = self._segment(blob)

        chunks = []

        channel_tags = ('mono', ) if len(channel_frames) == 1 else ('left', 'right')

        for frames, tag in zip(channel_frames, channel_tags):
            start = 0
            for idx, frame in enumerate(frames):
                chunks.append(dict(offset=idx, weight=1.0, blob=frame, location=[start, start + len(frame)], tags={'channel': tag}))
                start += self.hop_length

        return chunks
```

**Review by nan-wang — CHANGES_REQUESTED** [review]
I see your points. I've added some feedbacks

**Comment by shakurshams** [comment]
> I see your points. I've added some feedbacks
 
@nan-wang I like the idea of the tags. All feedbacks addressed. Need to run the tests.

**Review by JoanFM — COMMENTED** [review]
I would suggest to break this PR into smaller pieces to be able to discuss in a better way

**Comment by shakurshams** [comment]
each commit changes different  segmenters separately. we can make different PR from them. What do you think?

**Comment by JoanFM** [comment]
> each commit changes different segmenters separately. we can make different PR from them. What do you think?

that'd be great

**Comment by shakurshams** [comment]
@JoanFM @nan-wang I have created 5 separate PRs (#7981 #7982 #7983 #7984 #7985). Please check. We can close this one.

**Comment by nan-wang** [comment]
Splitted into #7981 #7982 #7983 #7984 #7985
